rfcs/002-merge-types-terms-syntax.md

```diff
@@ -0,0 +1,287 @@
+   something that is a type in the current syntax must be the corresponding
+   contract.
+
+### `term`
+
+As explained, we see the `term` translation as inserting implicit `§`
+automatically.
+
+- For most term-only constructs (that is everything excepted
+  records and constructors coming from types), `term` is an homomorphism (e.g.
+  for lists `term [t1, ..., tn] = [ term(t1), ..., term(tn)]`).
+- For type-only constructs (ground types, `->`, and `{_: T}`), it is
+  `term(T) = §(type(T))`, that is the contract associated to the type.
+
+#### Records
+
+It remains to treat the case of records, that can mix previously type and term
+constructs.
+
+- a row declaration `foo : Num` is translated as it is. This is actually already
+  supported since the piecewise signature feature. `{foo : Num}` is just
+  equivalent to `{foo | Num}` operationally when `foo` doesn't have a
+  definition.
+- For a tail, we have several solution:
+    1. Any record with a tail is first translated to a type, and then converted
+       to a contract. This follows mechanically the original principle, but
+       means the result is not a record: `{foo | Num; a}` would be elaborated to
+       a function. That is, `term(r with tail) = §type(r with tail)`.
+    2. More in the spirit of actually mixing terms and types, we could have
+       records to store their tail in a specific attribute, and adapt contract
+       application to take it into account. But this implies giving semantics to
+       things like `record.is_empty { ; a}` or `{ ; a} & { ; b}`. Those are
+       actually interesting questions outside of the context of this RFC (see
+       [201](https://github.com/tweag/nickel/issues/201)).
+  Solution 2. sounds more uniform (representing everything as records), but it
+  requires more design thinking. There isn't any obvious practical usage that
+  seems to require it right now (in practice we expect a record with a tail to be only used as a type annotation or applied as a contract, but nothing else).
+  Finally, we can always switch to 2. later, as it should be a
+  backward-compatible improvement.
+
+  For all those reasons, **we propose option 1. in this RFC**.
+
+### `type`
+
+Dually to terms:
+
+- For type-only constructs, `type` is the obvious homomorphism. For example,
+  `type(t -> t') = type(t) -> type(t')`.
+- For term-only constructs, we take the opaque type associated to a contract.
+  Concretely, we just translate it as a `Term`, and put a `#` in front:
+  `type(fun x => body)` = `#term(fun x => body)`.
+
+#### Records
+
+As for terms, the case of records remain. For records that are actually the
+embedding of a `Type`, such as `{foo : Num, bar: {baz: Num}}`, the translation
+is obvious. For records that are the embedding of a `Term`, we can simply
+proceed as above: `type({foo = 1, bar = "baz"}) = #{foo = 1, bar = "baz"}`.
+
+The case for records that mixes the both `Term` and `Type` is less clear. Take
```

Comment:
Would it make sense to consider `foo: Num = 1` as equivalent to `foo = (1: Num)` and translate it as such? (Maybe that’s equivalent to your first proposal, but I’m not 100% sure)

Reply:
Yes, I think it's equivalent.At any rate `{foo : Num = 1}` is actually desugared to `{foo = (1 : Num)}` by the parser, and this proposals says than then this is a term and the corresponding type will be `#{foo = (1 : Num)}`.